compiler-core/src/exhaustiveness.rs

```diff
@@ -29,208 +75,629 @@
+    ///       branch we know it will always match and we can produce a leaf in the
+    ///       decision tree (there's an exception when we have guards, but we'll
+    ///       get to it later)!
+    ///
+    ///     let otherwise = a
+    ///     ─┬───────────────
+    ///      ╰── And now we can understand what those `bindings` at the start of
+    ///          a body are: as we remove variable patterns, we will rewrite those
+    ///          as assignments at the top of the body of the corresponding branch.
+    ///
+    ///     otherwise
+    ///   }
+    /// }
+    /// ```
+    ///
+    fn move_unconditional_patterns(&mut self, compiler: &Compiler<'_>) {
+        self.checks.retain_mut(|check| {
+            loop {
+                match compiler.pattern(check.pattern) {
+                    // Variable patterns always match, so we move those to the body
+                    // and remove them from the branch's checks.
+                    Pattern::Variable { name } => {
+                        self.body.assign(name.clone(), check.var.clone());
+                        return false;
+                    }
+                    // A discard pattern always matches, but since the value is not
+                    // used we can just remove it without even adding an assignment
+                    // to the body!
+                    Pattern::Discard => return false,
+                    // Assigns are kind of special: they get turned into assignments
+                    // (shocking) but then we can't discard the pattern they wrap.
+                    // So we replace the assignment pattern with the one it's wrapping
+                    // and try again.
+                    Pattern::Assign { name, pattern } => {
+                        self.body.assign(name.clone(), check.var.clone());
+                        check.pattern = *pattern;
+                        continue;
+                    }
+                    // All other patterns are not unconditional, so we just keep them.
+                    _ => return true,
+                }
+            }
+        });
+    }
+}
 
-/// The body of code to evaluate in case of a match.
+/// The body of a branch. It always starts with a series of variable assignments
+/// in the form: `let a = b`. As explained in `move_unconditional_patterns`' doc,
+/// each body starts with a series of assignments we keep track of as we're
+/// compiling each branch.
+///
 #[derive(Clone, Eq, PartialEq, Debug)]
 pub struct Body {
     /// Any variables to bind before running the code.
     ///
-    /// The tuples are in the form `(name, source)` (i.e `bla = source`).
+    /// The tuples are in the form `(name, value)`, so `(wibble, var)`
+    /// corresponds to `let wibble = var`.
+    ///
     bindings: Vec<(EcoString, Variable)>,
 
     /// The index of the clause in the case expression that should be run.
-    clause_index: u16,
+    ///
+    branch_index: usize,
```

Comment:
Why is this called the branch index? It seems to still be the index of the clause, not a branch.

Reply:
Oops my bad!